codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java

```diff
@@ -349,17 +350,68 @@ public String doubleShape(DoubleShape shape) {
     @Override
     public String unionShape(UnionShape shape) {
         writer.addImportFromModule("github.com/dafny-lang/DafnyRuntimeGo", "dafny");
-        return """
-            func () types.%s {
-                var union types.%s
-                union = &types.MyUnionMemberIntegerValue{
-                    Value: %s.(int32),
-                }
-                return union
+        String functionSignature = "func() %s {".formatted(context.symbolProvider().toSymbol(shape));
+        String nilCheck = """
+                if %s == nil {
+                        return nil
+                }""".formatted(
+                        dataSource
+                );
+        if (GoPointableIndex.of(context.model()).isPointable(shape) == false) {
+            nilCheck = "";
+        }
+        String unionVariableInitialization = "var union %s".formatted(context.symbolProvider().toSymbol(shape));
+        String eachMemberInUnion = "";
+        for (var member : shape.getAllMembers().values()) {
+            Shape targetShape = context.model().expectShape(member.getTarget());
+            String memberName = context.symbolProvider().toMemberName(member);
+            String rawUnionDataSource = "(" + dataSource + ".(" + DafnyNameResolver.getDafnyType(shape, context.symbolProvider().toSymbol(shape)) + "))";
+            // unwrap union type, assert it then convert it to its member type with Dtor_ (example: Dtor_BlobValue()). unionDataSource is not a wrapper object until now.
```

Comment:
Even if it's Optional?

Reply:
Yes. In union we unwrap the membershape, assert it, convert to the required value and then again wrap it. This is because in other shapes we can directly do `dataSource.UnwrapOr(nil)` but to find a value inside dafny generated union we need to `Dtor_union()` then unwrap then assert and only then use Dtor once again (example `Dtor_MapValue()`). 

Example of string in 393 is `(dafnyInput.Dtor_union().UnwrapOr(nil).(simpleunioninternaldafnytypes.MyUnion)).Dtor_MapValue()`